### ISSUE TITLE
Unable to link cvxpy to SCIP

### ISSUE THREAD

**Issue by Mark531** [cvxpy/cvxpy#1889]
**Describe the bug**
I need to solve a non-linear MIP problem, so I installed pyscipopt. Though, after importing the package, cvxpy still cannot find the SCIP solver.

**To Reproduce**
```
import pyscipopt
import numpy as np
import cvxpy as cp

x = np.arange(100)
y = 5 * x**2 - 3 * x + 4
v = cp.Variable(integer=True)
problem = cp.Problem(cp.Minimize(cp.sum_squares(y - v)))
problem.solve(solver=cp.SCIP)
```

**Expected behavior**
The problem should be solved by SCIP.

**Output**
> SolverError: 
> 
>                     CVXPY needs additional software (a `mixed-integer solver`) to handle this model.
>                     The web documentation
>                         https://www.cvxpy.org/tutorial/advanced/index.html#mixed-integer-programs
>                     reviews open-source and commercial options for mixed-integer solvers.
> 
>                     Quick fix: if you install the python package CVXOPT (pip install cvxopt),
>                     then CVXPY can use the open-source mixed-integer solver `GLPK`.

**Version**
 - OS: Windows 10 64-bits
 - CVXPY Version: 1.1.3

**Comment by rileyjmurray** [comment]
What version of pyscipopt are you using? Per our documentation, we don't support pyscipopt 4 or higher: https://www.cvxpy.org/install/index.html#install-with-scip-support. (Our reason for this is that there is a correctness bug in SCIP 8, which is the only version of SCIP available with pyscipopt.

**Comment by Mark531** [comment]
You're right, I have installed the 4.2.0 version. I will downgrade the package, I just hope we won't lose important bug corrections...

**Comment by Mark531** [comment]
So, I installed the version 3.5.0 of pyscipopt and running the program above gives the very same error...

**Comment by SteveDiamond** [comment]
This is the code being run by CVXPY to verify that pyscipopt is installed:
```
from cvxpy.utilities.versioning import Version
import pyscipopt
v = pyscipopt.__version__
if Version(v) >= Version('4.0.0'):
    msg = 'PySCIPOpt (SCIP\'s Python wrapper) is installed and its ' \
          'version is %s. CVXPY only supports PySCIPOpt < 4.0.0.' % v
    raise ModuleNotFoundError(msg)
```

**Comment by Mark531** [comment]
Running your code triggers this error:

> ModuleNotFoundError: No module named 'cvxpy.utilities.versioning'

**Comment by SteveDiamond** [comment]
Can you try cvxpy 1.2?

**Comment by Mark531** [comment]
I upgraded cvxpy to 1.2.1 but when importing it, I got the following errors:

> (CVXPY) Sep 21 12:07:23 AM: Encountered unexpected exception importing solver CVXOPT:
> ImportError('DLL load failed: The specified module could not be found.')
> (CVXPY) Sep 21 12:07:23 AM: Encountered unexpected exception importing solver GLPK:
> ImportError('DLL load failed: The specified module could not be found.')
> (CVXPY) Sep 21 12:07:23 AM: Encountered unexpected exception importing solver GLPK_MI:
> ImportError('DLL load failed: The specified module could not be found.')

Is it expected?

**Comment by SteveDiamond** [comment]
That's certainly not expected, and doesn't happen in the CI (which tests Python 3.7-3.10 on Ubuntu, Windows, and Mac). It must be an issue with your environment. Are you using virtual environments?

**Comment by Mark531** [comment]
Yes, I installed it within an Anaconda virtual environment. I uninstalled cvxpy and reinstalled it, I got the same errors. Which versions of python have you tested your installs with?

**Comment by SteveDiamond** [comment]
3.7, 3.8, 3.9 and 3.10. You can see all the environments tested here:
https://github.com/cvxpy/cvxpy/actions/runs/3079069674

**Comment by Mark531** [comment]
Mmm, during the installation, these warnings popped up:

> WARNING: Ignoring invalid distribution -vxpy (c:\users\mark4\anaconda3\lib\site-packages)

Maybe there's something to investigate there.

**Comment by Mark531** [comment]
From your build pipeline (windows-2019, 3.7), you install cvxpy v1.3.0:

> Successfully installed cvxpy-1.3.0

Though the latest version available seems to be 1.2.1:

> (base) C:\Users\Mark4>pip index versions cvxpy
> WARNING: pip index is currently an experimental command. It may be removed/changed in a future release without prior warning.
> cvxpy (1.2.1)

How come?

**Comment by phschiele** [comment]
@Mark531 As soon as 1.2 is released, the master branch is bumped to 1.3, such that it starts accumulating features for the next release. That's common practice. The link just shows that we do test 3.7-3.10, which is also true for the 1.2.1 version.

**Comment by Mark531** [comment]
Well, this seems to show that the installation procedure didn't work well:

> (base) C:\Users\Mark4>pip install cvxpy
> Collecting cvxpy
>   Using cached cvxpy-1.2.1-cp37-cp37m-win_amd64.whl (833 kB)
> Requirement already satisfied: osqp>=0.4.1 in c:\users\mark4\anaconda3\lib\site-packages (from cvxpy) (0.6.1)
> Requirement already satisfied: scipy>=1.1.0 in c:\users\mark4\anaconda3\lib\site-packages (from cvxpy) (1.7.1)
> Requirement already satisfied: scs>=1.1.6 in c:\users\mark4\anaconda3\lib\site-packages (from cvxpy) (2.1.2)
> Requirement already satisfied: ecos>=2 in c:\users\mark4\anaconda3\lib\site-packages (from cvxpy) (2.0.7.post1)
> Requirement already satisfied: numpy>=1.15 in c:\users\mark4\anaconda3\lib\site-packages (from cvxpy) (1.20.3)
> Requirement already satisfied: future in c:\users\mark4\anaconda3\lib\site-packages (from osqp>=0.4.1->cvxpy) (0.18.2)
> Installing collected packages: cvxpy
> Successfully installed cvxpy-1.2.1
> 
> (base) C:\Users\Mark4>python
> Python 3.7.7 (default, May  6 2020, 11:45:54) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import cvxpy
> (CVXPY) Sep 21 12:47:09 AM: Encountered unexpected exception importing solver CVXOPT:
> ImportError('DLL load failed: The specified module could not be found.')
> (CVXPY) Sep 21 12:47:09 AM: Encountered unexpected exception importing solver GLPK:
> ImportError('DLL load failed: The specified module could not be found.')
> (CVXPY) Sep 21 12:47:09 AM: Encountered unexpected exception importing solver GLPK_MI:
> ImportError('DLL load failed: The specified module could not be found.')

**Comment by SteveDiamond** [comment]
Have you tried uninstalling cvxopt and reinstalling it?

**Comment by Mark531** [comment]
I've just reinstalled cvxopt and it did the trick! Now, cvxpy could be imported again.

So, to go back to the initial issue, the code now runs... but it never ends. I added a strong constraint about v to make the problem simpler:
`problem = cp.Problem(cp.Minimize(cp.sum_squares(y - v)), [v >= 3, v <= 4])`

But it runs endlessly. Could you have a try?

**Comment by SteveDiamond** [comment]
I ran it with ``problem.solve(solver=cp.SCIP, verbose=True)``. You can see SCIP is running, it's just taking forever as you said. But it's a SCIP issue at this point.

**Comment by Mark531** [comment]
Mmm, it's not normal to run forever to solve such a toy problem. When v is defined to be integer and between 3 and 4, it can only take 2 values! So, all the optimization engine has to do is try each value, compute the objective function and take the mimimum: it should complete instantaneously. There must be a bug somewhere, either in the cvxpy to SCIP problem transcription, or in SCIP itself.

**Comment by SteveDiamond** [comment]
We are looking into this more deeply and will have a new SCIP interface soonish that works with the latest SCIP/pyscipopt. We figured out that if we don't try and get dual variables for SOCPs then we can get correct results from pyscipopt 4.2. Even with the new interface we are still getting weird results on your toy problem though. This seems like a SCIP issue but will get confirmation from the SCIP devs (https://github.com/scipopt/scip/issues/22).

**Comment by Mark531** [comment]
Thank you very much!

**Comment by rileyjmurray** [comment]
@Mark531 the problem is just that your example is very badly scaled. It seems that the bad scaling is interacting with SCIP's presolve in a way that makes the continuous relaxations impossible to solve accurately. Unfortunately, it also seems that we need to leave SCIP's presolve on or else we can get incorrect results for problems with quadratic constraints. I've pointed this out to SCIP's developers so hopefully they can investigate if they have a bug in solving quadratic problems *without* presolve. But for the time being our plan is simply to leave presolve on by default for quadratic problems.

**Comment by Mark531** [comment]
What do you mean by "badly scaled"? Is there something I can do to rewrite the problem to help SCIP's "presolve" step?
I was also thinking... Maybe I should call SCIP directly instead of going with cvxpy. I don't know about the complexity of this solution.

**Comment by SteveDiamond** [comment]
"Badly scaled" means in this case that the dynamic range of values in the problem data is quite large, from ~1 to 5e4. Ideally problem data is centered around +/- 1. Most solvers will rescale the data to lie in the desired range but maybe SCIP isn't doing this. 

In any case I should have been more clear that we have a version of CVXPY ready for you that works with pyscipopt 4.2: https://github.com/cvxpy/cvxpy/tree/feature/scip4 (clone the branch and install from source).
As we said there's still issues on your toy problem, but I assume this is not what you actually want to use SCIP for. Calling SCIP directly would not have any advantage over using CVXPY, because the ultimate input to SCIP would be the same.

**Comment by rileyjmurray** [comment]
@Mark531 if you really do mean to solve a problem that looks like that, then you can at least replace the objective ``cp.Minimize(cp.sum_squares(expr))`` with ``cp.Minimize(cp.norm(expr))``. Steven pointed out that your numbers range 5 orders of magnitude, but on its own that isn't so bad. The numerical problems arise when you have numbers spanning 10 orders of magnitude (as happens with the ``cp.sum_squares`` formulation).

**Comment by Mark531** [comment]
Oh ok, I wouldn't have imagined that the range of x values would pose a problem, I thought that the complexity of the problem was rather driven by the range of v values (i.e. finite set versus infinite set).

As you pointed it out, it was just a toy problem, the actual problem I'm looking at currently deals with much smaller numbers. I'll check it then.

**Comment by rileyjmurray** [comment]
The practical problem with large numbers is that computers work in finite-precision arithmetic. This introduces rounding errors into operations as innocuous as addition, multiplication, division, and subtraction. Such rounding errors can cause very reliable algorithms to fail if a problem is badly scaled.

There is also a fun fact about convex optimization theory on this topic: the runtime of provably-correct algorithms for convex optimization -- even linear programming -- depend on the number of bits required to represent a problem. So if you have a problem that requires accurate representation of numbers spanning huge orders of magnitude then even the best algorithms would be sensitive to problem scale. (This isn't relevant in practice since practically all numerical solvers are designed to use IEEE double precision arithmetic.)

**Comment by Transurgeon** [comment]
closing this issue since it seemed like a problem instance specific issue with scaling.
